proxy/src/transparency.rs

```diff
@@ -0,0 +1,688 @@
+
+        Protocol::Tcp
+    }
+}
+
+// ===== impl HttpBody =====
+
+impl tower_h2::Body for HttpBody {
+    type Data = Bytes;
+
+    fn is_end_stream(&self) -> bool {
+        match *self {
+            HttpBody::Http1(_) => false,
+            HttpBody::Http2(ref b) => b.is_end_stream(),
+        }
+    }
+
+    fn poll_data(&mut self) -> Poll<Option<Self::Data>, h2::Error> {
+        match *self {
+            HttpBody::Http1(ref mut b) => {
+                match b.poll() {
+                    Ok(Async::Ready(Some(chunk))) => Ok(Async::Ready(Some(chunk.into()))),
+                    Ok(Async::Ready(None)) => Ok(Async::Ready(None)),
+                    Ok(Async::NotReady) => Ok(Async::NotReady),
+                    Err(e) => {
+                        debug!("http/1 body error: {}", e);
+                        Err(h2::Reason::INTERNAL_ERROR.into())
+                    }
+                }
+            },
+            HttpBody::Http2(ref mut b) => b.poll_data().map(|async| async.map(|opt| opt.map(|data| data.into()))),
+        }
+    }
+
+    fn poll_trailers(&mut self) -> Poll<Option<http::HeaderMap>, h2::Error> {
+        match *self {
+            HttpBody::Http1(_) => Ok(Async::Ready(None)),
+            HttpBody::Http2(ref mut b) => b.poll_trailers(),
+        }
+    }
+}
+
+impl Default for HttpBody {
+    fn default() -> HttpBody {
+        HttpBody::Http2(Default::default())
+    }
+}
+
+#[derive(Debug)]
+pub struct BodyStream<B>(B);
+
+impl<B> Stream for BodyStream<B>
+where
+    B: tower_h2::Body,
+{
+    type Item = BufAsRef<<B::Data as ::bytes::IntoBuf>::Buf>;
+    type Error = hyper::Error;
+
+    fn poll(&mut self) -> Poll<Option<Self::Item>, Self::Error> {
+        self.0.poll_data()
+            .map(|async| async.map(|opt| opt.map(|buf| BufAsRef(::bytes::IntoBuf::into_buf(buf)))))
+            .map_err(|e| {
+                trace!("h2 body error: {:?}", e);
+                hyper::Error::Io(io::ErrorKind::Other.into())
+            })
+    }
+}
+
+#[derive(Debug)]
+pub struct BufAsRef<B>(B);
+
+impl<B: ::bytes::Buf> AsRef<[u8]> for BufAsRef<B> {
+    fn as_ref(&self) -> &[u8] {
+        ::bytes::Buf::bytes(&self.0)
+    }
+}
+
+
+// ===== impl HyperServerSvc =====
+
+struct HyperServerSvc<S> {
+    service: RefCell<S>,
+    srv_ctx: Arc<ServerCtx>,
+}
+
+impl<S, B> hyper::server::Service for HyperServerSvc<S>
+where
+    S: Service<
+        Request=http::Request<HttpBody>,
+        Response=http::Response<B>,
+    >,
+    S::Error: fmt::Debug,
+    S::Future: 'static,
+    B: tower_h2::Body + 'static,
+{
+    type Request = hyper::server::Request;
+    type Response = hyper::server::Response<BodyStream<B>>;
+    type Error = hyper::Error;
+    //TODO: unbox me
+    type Future = Box<Future<Item=Self::Response, Error=Self::Error>>;
+
+    fn call(&self, req: Self::Request) -> Self::Future {
+        let mut req: http::Request<hyper::Body> = req.into();
+        req.extensions_mut().insert(self.srv_ctx.clone());
+
```

Comment:
Some HTTP/1-specific proxy stuff happens in here, perhaps it'd be better to lift this higher up so as to not seem like simply hyper glue...